virtcontainers/persist/fs/fs.go

```diff
@@ -0,0 +1,277 @@
+// Copyright (c) 2016 Intel Corporation
+// Copyright (c) 2018 Huawei Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package fs
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	persistapi "github.com/kata-containers/runtime/virtcontainers/persist/api"
+	"github.com/sirupsen/logrus"
+)
+
+// persistFile is the file name for JSON sandbox/container configuration
+const persistFile = "persist.json"
+
+// dirMode is the permission bits used for creating a directory
+const dirMode = os.FileMode(0700)
+
+// fileMode is the permission bits used for creating a file
+const fileMode = os.FileMode(0640)
+
+// storagePathSuffix is the suffix used for all storage paths
+//
+// Note: this very brief path represents "virtcontainers". It is as
+// terse as possible to minimise path length.
+const storagePathSuffix = "vc"
+
+// sandboxPathSuffix is the suffix used for sandbox storage
+const sandboxPathSuffix = "sbs"
+
+// runStoragePath is the sandbox runtime directory.
+// It will contain one state.json and one lock file for each created sandbox.
+var runStoragePath = filepath.Join("/run", storagePathSuffix, sandboxPathSuffix)
+
+// FS storage driver implementation
+type FS struct {
+	sandboxState   *persistapi.SandboxState
+	containerState map[string]persistapi.ContainerState
+	setFuncs       map[string]persistapi.SetFunc
+
+	lockFile *os.File
+}
+
+var fsLog = logrus.WithField("source", "virtcontainers/persist/fs")
+
+// Logger returns a logrus logger appropriate for logging Store messages
+func (fs *FS) Logger() *logrus.Entry {
+	return fsLog.WithFields(logrus.Fields{
+		"subsystem": "persist",
+	})
+}
+
+// Name returns driver name
+func Name() string {
+	return "fs"
+}
+
+// Init FS persist driver and return abstract PersistDriver
+func Init() (persistapi.PersistDriver, error) {
+	return &FS{
+		sandboxState:   &persistapi.SandboxState{},
+		containerState: make(map[string]persistapi.ContainerState),
+		setFuncs:       make(map[string]persistapi.SetFunc),
+	}, nil
+}
+
+func (fs *FS) sandboxDir() (string, error) {
+	id := fs.sandboxState.SandboxContainer
+	if id == "" {
+		return "", fmt.Errorf("sandbox container id required")
+	}
+
+	return filepath.Join(runStoragePath, id), nil
+}
+
+// ToDisk sandboxState and containerState to disk
+func (fs *FS) ToDisk() (retErr error) {
+	// call registered hooks to set sandboxState and containerState
+	for _, fun := range fs.setFuncs {
+		fun(fs.sandboxState, fs.containerState)
+	}
+
+	sandboxDir, err := fs.sandboxDir()
+	if err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(sandboxDir, dirMode); err != nil {
+		return err
+	}
+
+	if err := fs.lock(); err != nil {
+		if err1 := fs.Destroy(); err1 != nil {
```

Comment:
Nothing is written yet. I think we should not `Destroy()` existing files, otherwise it can be dangerous if concurrent access is possible.